### PR TITLE
Upstream MIPS pmap fixes

### PR DESCRIPTION
**Pull request by nwf**
Address race conditions between the MIPS pmap and TLB fault paths.

### CODE REVIEW

File: sys/mips/mips/pmap.c [hunk 1]
```diff
@@ -1727,6 +1745,22 @@ pmap_try_insert_pv_entry(pmap_t pmap, vm_page_t mpte, vm_offset_t va,
 
 /*
  * pmap_remove_pte: do the things to unmap a page in a process
+ *
+ * Returns true if this was the last pte in the pt (and possibly the last pt in
```
Comment (jrtc27) [inline sys/mips/mips/pmap.c:1749]: Please consistently capitalise PTE/PT/PD/PDE (unless specifically referring to variables by name).

File: sys/mips/mips/pmap.c [hunk 2]
```diff
@@ -1039,12 +1047,22 @@ _pmap_unwire_ptp(pmap_t pmap, vm_offset_t va, vm_page_t m)
...
 		/*
-		 * Recursively decrement next level pagetable refcount
+		 * Recursively decrement next level pagetable refcount.
+		 * Either that shoots down a larger range from TLBs (below)
+		 * or we're to shoot down just the page in question.
 		 */
 		pdp = (pd_entry_t *)*pmap_segmap(pmap, va);
 		pdpg = PHYS_TO_VM_PAGE(MIPS_DIRECT_TO_PHYS(pdp));
-		pmap_unwire_ptp(pmap, va, pdpg);
+		if(!pmap_unwire_ptp(pmap, va, pdpg)) {
```
Comment (jrtc27) [inline sys/mips/mips/pmap.c:1056]:
```suggestion
		if (!pmap_unwire_ptp(pmap, va, pdpg)) {
```

File: sys/mips/mips/pmap.c [hunk 3]
```diff
@@ -1795,8 +1829,12 @@ pmap_remove_page(struct pmap *pmap, vm_offset_t va)
 	if (!pte_test(ptq, PTE_V))
 		return;
 
-	(void)pmap_remove_pte(pmap, ptq, va, *pde);
-	pmap_invalidate_page(pmap, va);
+	/*
+	 * Remove this PTE from the PT.  If this is the last one, then
+	 * the TLB has already been shot down, so don't bother again
+	 */
+	if(!pmap_remove_pte(pmap, ptq, va, *pde))
```
Comment (jrtc27) [inline sys/mips/mips/pmap.c:1836]:
```suggestion
	if (!pmap_remove_pte(pmap, ptq, va, *pde))
```

File: sys/mips/mips/pmap.c [hunk 4]
```diff
@@ -1943,10 +1996,11 @@ pmap_remove_all(vm_page_t m)
 			    __func__, (void *)pv->pv_va, (uintmax_t)tpte));
 			vm_page_dirty(m);
 		}
-		pmap_invalidate_page(pmap, pv->pv_va);
+
+		if(!pmap_unuse_pt(pmap, pv->pv_va, *pde))
```
Comment (jrtc27) [inline sys/mips/mips/pmap.c:2000]:
```suggestion
		if (!pmap_unuse_pt(pmap, pv->pv_va, *pde))
```

File: sys/mips/mips/pmap.c (hunk 1 above)
Comment (nwf) [reply sys/mips/mips/pmap.c]: Good call; that's what comes of spreading the work across two days, I suppose.  If you see any other inconsistencies still lingering, please advise.

**Comment by nwf** [comment]
These have been merged upstream; see 1d5430c5750a5bfddec79000ce12260037e9eeee and 4a4e2a08fd6bfb55db1ea7edec8b334b9f77e827 .

**Pull request closed without merging**